### PR TITLE
Source SmartSheets: fix oauth2 implementation

### PR DESCRIPTION
**Pull request by artem1205**
## What
Resolving https://github.com/airbytehq/oncall/issues/1127

## How

Fix refresh token by using SingleUseRefreshTokenOauth2Authenticator

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
reauthenticate will be needed for all users

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

### CODE REVIEW

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml [hunk 1]
```diff
@@ -2,9 +2,13 @@ connector_image: airbyte/source-smartsheets:dev
 tests:
   spec:
     - spec_path: "source_smartsheets/spec.json"
+      backward_compatibility_tests_config:
+        disable_for_version: "0.1.14"
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml:6]: Can you share more about what the user experience will be with this change?  If they will need to re-authenticate, this is a breaking change and should be `v1.0.0`

Comment (artem1205) [reply airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml:6]: Re-authentication will be needed for all users, as spec file structure will be changed, should I bump `version` to `1.0.0`?

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json [hunk 2]
```diff
@@ -4,15 +4,68 @@
...
+            "title": "OAuth2.0",
+            "required": ["client_id", "client_secret", "refresh_token", "access_token", "token_expiry_date"],
+            "properties": {
+              "auth_type": {
+                "type": "string",
+                "const": "oauth2.0"
+              },
+              "client_id": {
+                "type": "string",
+                "description": "The API ID of the Gitlab developer application.",
```
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json:26]:
```suggestion
                "description": "The API ID of the SmartSheets developer application.",
```

There are a few references to GitLab in this file

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java [hunk 3]
```diff
@@ -23,14 +24,17 @@ public class SmartsheetsOAuthFlow extends BaseOAuth2Flow {
...
   private static final String ACCESS_TOKEN_URL = "https://api.smartsheet.com/2.0/token";
+  private final Clock clock;
 
   public SmartsheetsOAuthFlow(final ConfigRepository configRepository, final HttpClient httpClient) {
     super(configRepository, httpClient);
+    this.clock = Clock.systemUTC();
   }
 
   @VisibleForTesting
   public SmartsheetsOAuthFlow(final ConfigRepository configRepository, final HttpClient httpClient, final Supplier<String> stateSupplier) {
```
Comment (pedroslopez) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java:35]: Note that oauth flows are being removed from this repo. These changes will have to be done on the airbyte-platform repo's airbyte-oauth module

File: airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml (hunk 1 above)
Comment (pedroslopez) [reply airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml:6]: If that's the case then yes, this should be a breaking change and need to be communicated to users.

However, in other oauth implementations we have maintained compatibility with older configs by still looking for an access token in the old path so we don't break existing connections. Could we do something similar here?

**Comment by artem1205** [comment]
@pedroslopez , unfortunately not, as we need extra data for [reauthentication](https://github.com/airbytehq/airbyte/pull/23237/files#diff-acfdd07fe32bdbdc73b8c93c72d5270daa64959bd1ff9b70a49792ad9181747aR26), that we don't have now in our DB.

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json (hunk 2 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json]: updated

File: airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml (hunk 1 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-smartsheets/acceptance-test-config.yml:6]: version updated

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java (hunk 3 above)
Comment (artem1205) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java:35]: PR created: https://github.com/airbytehq/airbyte-platform/pull/41

**Comment by artem1205** [comment]
Hi, @jnr0790 
As this PR contains breaking changes, I would ask you to communicate with customers so that they can reauthenticate.

cc @lazebnyi @YowanR

**Review by evantahler — APPROVED** [review]
👍 and noting the breaking change

**Review by lazebnyi — APPROVED** [review]

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java (hunk 3 above)
Comment (pedroslopez) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/SmartsheetsOAuthFlow.java]: New PR in https://github.com/airbytehq/airbyte-platform-internal/pull/4813 👍🏻

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json [hunk 4]
```diff
@@ -4,15 +4,68 @@
...
+            "required": ["access_token"],
+            "properties": {
+              "auth_type": {
+                "type": "string",
+                "const": "access_token"
+              },
+              "access_token": {
+                "type": "string",
+                "title": "Private Token",
+                "description": "The access token to use for accessing your data from Smartsheets. This access token must be generated by a user with at least read access to the data you'd like to replicate. Generate an access token in the Smartsheets main menu by clicking Account > Apps & Integrations > API Access. See the <a href=\"https://docs.airbyte.com/integrations/sources/amplitude#setup-guide\">setup guide</a> for information on how to obtain this token.",
```
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json:63]: This link is to the amplitude connector setup guide!

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json [hunk 5]
```diff
@@ -4,15 +4,68 @@
...
+              },
+              "refresh_token": {
+                "type": "string",
+                "description": "The key to refresh the expired access_token.",
+                "airbyte_secret": true
+              }
+            }
+          },
+          {
+            "title": "Private Token",
```
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json:52]: It looks like Smartsheets calls these "API Access Token"s https://help.smartsheet.com/articles/2482389-generate-API-key

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json (hunk 4 above)
Comment (pedroslopez) [reply airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json:63]: We need to update the docs for this connector to reflect the new ways of setting it up. Not sure if you want to do it in this PR or if this is something for the docs team.

**Review by pedroslopez — CHANGES_REQUESTED** [review]
Code looks good, though did not functionally test this. Requested changes are for naming and an incorrect link.

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json (hunk 5 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json]: updated

File: airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json (hunk 4 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-smartsheets/source_smartsheets/spec.json]: updated

**Review by pedroslopez — APPROVED** [review]

**Pull request merged**